docs/protocol.md

```diff
@@ -0,0 +1,282 @@
+
+There are other types of messages denoted by `message_type`, but these are
+only used in the "bootstrapping" procedure when a node needs to synchronize
+itself with all the blocks in the distributed ledger (in a bulk fashion).
+Bootstrapping is done over TCP connections, not with UDP packets.
+
+All UDP messages have a fixed-size header (see `rai::message::read_header()` in `rai/node/common.cpp`):
+
+  | Field         | Size (bytes)  | Description |
+  | ------------- | ------------- | ----------- |
+  | Magic         | 2             | Magic number to indicate this is a Nano packet |
+  | VersionMax    | 1             | Maximum protocol version the node that sent the packet can understand |
+  | VersionUsing  | 1             | Protocol version the packet uses |
+  | VersionMin    | 1             | Minimum protocol version the node that sent the packet can understand |
+  | Type          | 1             | Message type |
+  | Extensions    | 2             | 16-bit (little-endian) value that indicates possible extensions, optionally contains a block type  |
+
+The 2-byte magic number has a value that depends on the type of Nano network on which
+the message is sent:
+
+  | Value | Network | Description |
+  | ----- | ------- | ----------- |
+  | `RA`  | Testing | Internal testing network used for development |
+  | `RB`  | Beta    | Beta network used for development |
+  | `RC`  | Live    | Live network, where real transaction in Nano are performed |
+
+Each of these 3 networks has their own genesis account, block and amount. Blocks
+intended for one network will not be accepted on another network, as they will
+fail to validate. XXX is this true?
+
+The VersionMax, VersionUsing and VersionMin values are related to the version
+of the network protocol being used. The VersionUsing value indicates the protocol
+version a packet is adhering to.
+
+The message Type can have one of the following values (see `message_type` in `rai/node/common.hpp`):
+
+  | Type       | Value |
+  | ---------- | ----- |
+  | Keepalive  | 2     |
+  | Publish    | 3     |
+  | ConfirmReq | 4     |
+  | ConfirmAck | 5     |
+
+The Extensions field currently only contains the 8-bit block type, in the highest
+8 bits of the field. But this is only the case if the message contains a Block (i.e. Publish, ConfirmReq and ConfirmAck messages).
+See `rai::message::block_type_set()` in `rai/node/common.cpp`.
+
+## Keepalive message
+
+Reference: class `rai::keepalive` in `rai/node/common.cpp`
+
+A Keepalive message contains a list of IP addresses of peers, in the form of (IPv6 address, port) pairs.
+Each IP address is stored as 16 bytes, each port as 2 bytes:
+
+  | Field        | Size (bytes) | Description     |
+  | ------------ | ------------ | --------------- |
+  | IPv6 address | 16           | Big-endian?     |
+  | Port         | 2            | Little-endian?! |
+
+The number of pairs in the message can vary, from 1 up to 8 peers.
```

Comment:
Evidently there's currently a bug where there must be 8 peers or the reference node does an out of bounds read https://github.com/nanocurrency/raiblocks/issues/673

Reply:
@termhn that's fixed

Reply:
Sweet